backend/fms_core/template_importer/row_handlers/project_link_samples/project_link_samples.py

```diff
@@ -28,18 +32,94 @@ def process_row_inner(self, sample, project, action):
                               f"is named {sample_obj.name} not {sample['sample_name']}."
                 self.warnings['sample'].append(warning_msg)
 
-            # Check if link exists to ensure there's not a duplicated association
-            if action['name'] == ADD_ACTION:
+            # Perform an add project and/or study action
+            if action['name'] == ADD_PROJECT_ACTION:
                 # Create link object if no errors
                 link_created, self.errors['link'], self.warnings['link'] = create_link(sample=sample_obj,
                                                                                        project=project_obj)
 
-            # If the link doesn't exists we can't perform a remove action
-            elif action['name'] == REMOVE_ACTION:
-                # Remove link object if no errors
-                link_removed, self.errors['link'], self.warnings['link'] = remove_link(sample=sample_obj,
-                                                                                       project=project_obj)
+                # Throw error if study letter is provided
+                if project['study_letter']:
+                    self.errors['add_to_project'] = f"A study was provided. If you want to add the sample to a study use the [Add to study] action."
+
+            
+            # Perform an add study action
+            elif action['name'] == ADD_STUDY_ACTION:
+                # Make sure the sample is already associated to the project of the given study 
+                for derived_sample in sample_obj.derived_samples.all():
+                    if derived_sample.project is not None and derived_sample.project.id != project_obj.id:
+                        self.errors['add_to_study'] = (f"[Sample {sample_obj.name}] is not associated to project [{project_obj.name}].")
```

Comment:
Not true. The pools are a complex case ... 

The way they will be dealt with as far as i see it they will be queued by study as long as a single sample is part of that study. So if we receive a ready to sequence pool. All samples in the pool will be part of a single study (or different study in the same project). We will queue de pool sample once for the whole study. For this case your test would work but most pools will be made "in house" and will contain a mix of different project samples). 

Naturally when the samples in a pool will pass the pool step the next step will use the same sample (pool) in each study they belong to. This will lead to multiple instance of the same sample_next_step being scheduled, we will ignore duplicates a single pool by study will support each sample queued to the workflow for that study. That sample derived sample will not all belong to the same project. 

So if we want to manually add a pool to the study we have to support this too. I would include is_pool to check if everything is ok. If you have a pool you can have sample that do not belong to the project but at least one sample as to belong to the project. Also if you have a pool and there is already an identical sample_next_step you may want to return a simple warning. 

Lastly these tests should be included in the queue_sample_to_study_workflow service instead since they likely apply to all cases we will add sample to studies.

Reply:
`Also if you have a pool and there is already an identical sample_next_step you may want to return a simple warning.`

Identical meaning (sample, study)?

Reply:
sample, study and order

Reply:
I can return a warning but wouldn't it be caught by the unique constraint [sample, study, order]?

Reply:
Sample1, Study1, None is considered different from Sample1, Study1, None but Sample1, Study1, Order1 is considered identical to Sample1, Study1, Order1. this is part of why it would be good to send a warning. I don't think this really apply to the row handler function since we would not want people to add sample at the end of the workflow, but these checks should be done in the service and at that level we would want that.
If i assign a sample to the end of the workflow and the sample already completed the workflow once we may not want a second copy of the sample_next_step identical (to us not to the DB) to the one already there. Also when pooling all samples will become a single sample for the next step so there will likely be duplicate sample_next_step after completing the pooling step, one way to do the next step would be to make a set to ensure we don't have duplicates but we could also call a service  and have it ignore duplicates with a warning. The ignore option is not really a good practice, but it might be faster and simpler.
So the tldr version is if you do the test in the row_handler the unique constraint validation should be sufficient and an error is possibly a good way to tell the user the sample is already queued there.